### ISSUE TITLE
Add an API to obtain all reachable items of a crate

### ISSUE THREAD

**Issue by momvart** [rust-lang/rust#121525]
`TyCtxt` currently provides an API to get all exported symbols of a crate. Still, there are some items in external crates that are not necessarily exported but are reachable and there is no direct way to get them. For example, some public generic functions provided by a library are like that.

Although it is possible to find them using methods like `module_children`, I want to request an API in `TyCtxt` that returns all reachable `DefId`s, a superset of `exported_symbols`.

**Comment by oli-obk** [comment]
Does https://doc.rust-lang.org/stable/nightly-rustc/rustc_middle/query/plumbing/struct.TyCtxtAt.html#method.collect_and_partition_mono_items contain everything you need?

**Comment by oli-obk** [comment]
If it's not needed in rustc and can be emulated with existing functions, I think this should be implemented in the tool that needs it.

**Comment by momvart** [comment]
> If it's not needed in rustc and can be emulated with existing functions, I think this should be implemented in the tool that needs it.

That's right. Maybe `smir` can provide it? Anyway, the data resides in `CrateMetadata` protected from external access. 

BTW, although it's just a printing function, but it looks like [`rustc_middle::ty::print::pretty::for_each_def`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/print/pretty/fn.for_each_def.html) is doing the recursion to obtain the same list.

**Comment by oli-obk** [comment]
If that function can be refactored into something reusable, great! Let's do that

**Comment by krishpranav** [comment]
Hi, is this still open. @ maintaners can you assign this to me so that i can work on the PR for this.

**Comment by krishpranav** [comment]
PR for this issue: https://github.com/rust-lang/rust/pull/135436
